core/services/registrysyncer/syncer.go

```diff
@@ -0,0 +1,250 @@
+					},
+				},
+			},
+		},
+	}
+
+	contractReaderConfigEncoded, err := json.Marshal(contractReaderConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	cr, err := relayer.NewContractReader(ctx, contractReaderConfigEncoded)
+	if err != nil {
+		return nil, err
+	}
+
+	err = cr.Bind(ctx, []types.BoundContract{
+		{
+			Address: remoteRegistryAddress,
+			Name:    "CapabilitiesRegistry",
+		},
+	})
+
+	return cr, err
+}
+
+func newSyncer(
+	stopCh services.StopChan,
+	lggr logger.Logger,
+	reader types.ContractReader,
+) *registrySyncer {
+	return &registrySyncer{
+		stopCh: stopCh,
+		lggr:   lggr,
+		reader: reader,
+	}
+}
+
+func (s *registrySyncer) Start(ctx context.Context) error {
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+		s.syncLoop()
+	}()
+	return nil
+}
+
+func (s *registrySyncer) syncLoop() {
+	ctx, cancel := s.stopCh.NewCtx()
+	defer cancel()
+
+	ticker := time.NewTicker(defaultTickInterval)
+	defer ticker.Stop()
+
+	// Sync for a first time outside the loop; this means we'll start a remote
+	// sync immediately once spinning up syncLoop, as by default a ticker will
+	// fire for the first time at T+N, where N is the interval.
+	s.lggr.Debug("starting initial sync with remote registry")
+	err := s.sync(ctx)
+	if err != nil {
+		s.lggr.Errorw("failed to sync with remote registry", "error", err)
+	}
+
+	for {
+		select {
+		case <-s.stopCh:
+			return
+		case <-ticker.C:
+			s.lggr.Debug("starting regular sync with the remote registry")
+			err := s.sync(ctx)
+			if err != nil {
+				s.lggr.Errorw("failed to sync with remote registry", "error", err)
+			}
+		}
+	}
+}
+
+func (s *registrySyncer) state(ctx context.Context) (State, error) {
+	dons := []kcr.CapabilitiesRegistryDONInfo{}
+	err := s.reader.GetLatestValue(ctx, "CapabilitiesRegistry", "getDONs", nil, &dons)
+	if err != nil {
+		return State{}, err
+	}
+
+	idsToDONs := map[DonID]kcr.CapabilitiesRegistryDONInfo{}
+	for _, d := range dons {
+		idsToDONs[DonID(d.Id)] = d
+	}
+
+	caps := []kcr.CapabilitiesRegistryCapabilityInfo{}
+	err = s.reader.GetLatestValue(ctx, "CapabilitiesRegistry", "getCapabilities", nil, &caps)
+	if err != nil {
+		return State{}, err
+	}
+
+	idsToCapabilities := map[HashedCapabilityID]kcr.CapabilitiesRegistryCapabilityInfo{}
+	for _, c := range caps {
+		idsToCapabilities[c.HashedId] = c
+	}
+
+	nodes := []kcr.CapabilitiesRegistryNodeInfo{}
+	err = s.reader.GetLatestValue(ctx, "CapabilitiesRegistry", "getNodes", nil, &nodes)
+	if err != nil {
+		return State{}, err
+	}
+
+	idsToNodes := map[p2ptypes.PeerID]kcr.CapabilitiesRegistryNodeInfo{}
+	for _, node := range nodes {
+		idsToNodes[node.P2pId] = node
+	}
+
+	return State{IDsToDONs: idsToDONs, IDsToCapabilities: idsToCapabilities, IDsToNodes: idsToNodes}, nil
+}
+
+func (s *registrySyncer) sync(ctx context.Context) error {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	if len(s.launchers) == 0 {
+		s.lggr.Warn("sync called, but no launchers are registered; nooping")
+		return nil
+	}
```

Comment:
I like this. I would even take this a step further and not start the sync loop if there are no launchers (I'm assuming these aren't going to be added dynamically anyway). It seems like an error to me if we have the registry with no listeners 🤷‍♂️